src/lib/datatypes/sol-memdesc.c

```diff
@@ -0,0 +1,804 @@
+    if (r < 0)
+        return r;
+
+    SOL_MEMDESC_FOREACH_ARRAY_ELEMENT_IN_RANGE(desc, src_memory, 0, len, idx, src_item) {
+        void *dst_item = sol_memdesc_get_array_element(desc, dst_memory, idx);
+
+        if (!dst_item) {
+            r = -errno;
+            goto failure;
+        }
+
+        r = set_content(desc->array_item, dst_item, src_item);
+        if (r < 0)
+            goto failure;
+    }
+
+    if (idx < len) {
+        r = errno ? -errno : -EINVAL;
+        goto failure;
+    }
+
+    return 0;
+
+failure:
+    sol_memdesc_resize_array(desc, dst_memory, idx);
+    return r;
+}
+
+static int
+compare_content(const struct sol_memdesc *desc, const void *a_mem, const void *b_mem)
+{
+    if (desc->ops && desc->ops->compare)
+        return desc->ops->compare(desc, a_mem, b_mem);
+
+#define RET_CMP_INT(type) \
+    do { \
+        const type *a = a_mem; \
+        const type *b = b_mem; \
+        if (*a < *b) \
+            return -1; \
+        else if (*a > *b) \
+            return 1; \
+        else \
+            return 0; \
+    } while (0)
+
+    switch (desc->type) {
+    case SOL_MEMDESC_TYPE_UINT8:
+        RET_CMP_INT(uint8_t);
+    case SOL_MEMDESC_TYPE_UINT16:
+        RET_CMP_INT(uint16_t);
+    case SOL_MEMDESC_TYPE_UINT32:
+        RET_CMP_INT(uint32_t);
+    case SOL_MEMDESC_TYPE_UINT64:
+        RET_CMP_INT(uint64_t);
+    case SOL_MEMDESC_TYPE_ULONG:
+        RET_CMP_INT(unsigned long);
+    case SOL_MEMDESC_TYPE_SIZE:
+        RET_CMP_INT(size_t);
+    case SOL_MEMDESC_TYPE_INT8:
+        RET_CMP_INT(int8_t);
+    case SOL_MEMDESC_TYPE_INT16:
+        RET_CMP_INT(int16_t);
+    case SOL_MEMDESC_TYPE_INT32:
+        RET_CMP_INT(int32_t);
+    case SOL_MEMDESC_TYPE_INT64:
+        RET_CMP_INT(int64_t);
+    case SOL_MEMDESC_TYPE_LONG:
+        RET_CMP_INT(long);
+    case SOL_MEMDESC_TYPE_SSIZE:
+        RET_CMP_INT(ssize_t);
+    case SOL_MEMDESC_TYPE_BOOLEAN: {
+        const bool *a = a_mem;
+        const bool *b = b_mem;
+
+        if (!*a && *b)
+            return -1;
+        else if (*a && !*b)
+            return 1;
+        else
+            return 0;
+    }
+    case SOL_MEMDESC_TYPE_DOUBLE: {
+        const double *a = a_mem;
+        const double *b = b_mem;
+
+        if (sol_util_double_equal(*a, *b))
+            return 0;
+        else if (*a < *b)
+            return -1;
+        else
+            return 1;
+    }
+    case SOL_MEMDESC_TYPE_CONST_STRING:
+    case SOL_MEMDESC_TYPE_STRING: {
+        const char *const *a = a_mem;
+        const char *const *b = b_mem;
+
+        if (!*a && *b)
+            return -1;
+        else if (*a && !*b)
+            return 1;
+        else if (*a == *b)
+            return 0;
+        else
+            return strcmp(*a, *b);
+    }
+    case SOL_MEMDESC_TYPE_PTR: {
+        const void *const *a = a_mem;
+        const void *const *b = b_mem;
+
+        if (!desc->pointed_item || !*a || !*b) {
+            if (*a < *b)
```

Comment:
makes sense. I'll keep NULL x some-pointer in consideration, but no less/greater then as it's not stable.